application/dash_application/views/bcl2barcode.py

```diff
@@ -61,74 +136,100 @@ def trim(row, index_name, len_name):
+# Assign the expected Pinery indices to bcl2barcode
+# If Pinery does not have that Run/Lane it will be excluded here
 bcl2barcode = bcl2barcode.merge(
-    min_index1_length,
-    how='left',
-    left_on=['Run Alias', 'Lane Number'],
-    right_on=['sequencerRunName', 'laneNumber']
+    index1_expected, how='inner',
+    left_on=[bcl2barcode_col.Run, bcl2barcode_col.Lane],
+    right_on=[PINERY_COL.SequencerRunName, PINERY_COL.LaneNumber],
+    suffixes=('', '_known')
 )
 bcl2barcode = bcl2barcode.merge(
-    min_index2_length,
-    how='left',
-    left_on=['Run Alias', 'Lane Number'],
-    right_on=['sequencerRunName', 'laneNumber'],
-    suffixes=['', '_del']
-)
-bcl2barcode = bcl2barcode.drop(['sequencerRunName_del', 'laneNumber_del'], axis=1)
-
-bcl2barcode['Index1 Trimmed'] = bcl2barcode.apply(lambda x: trim(x, 'Index1', 'Index1 min length'), axis=1)
-bcl2barcode['Index2 Trimmed'] = bcl2barcode.apply(lambda x: trim(x, 'Index2', 'Index2 min length'), axis=1)
-
-# Get single and dual index info using trimmed index
-pinery_single_index = pinery_with_expanded_barcodes[pinery_with_expanded_barcodes['Index2'].isna()]
-single_index = bcl2barcode.merge(
-    pinery_single_index, how='left',
-    right_on=['sequencerRunName', 'laneNumber', 'Index1 Trimmed'],
-    left_on=['sequencerRunName', 'laneNumber', 'Index1 Trimmed'],
-    suffixes=('', "_pinery")
+    index2_expected, how='inner',
+    left_on=[PINERY_COL.SequencerRunName, PINERY_COL.LaneNumber],
+    right_on=[PINERY_COL.SequencerRunName, PINERY_COL.LaneNumber],
+    suffixes=('', '_known')
 )
 
-# Unknown barcodes will have all Pinery columns NaN. Picked study title to check
-known = single_index[~single_index['studyTitle'].isna()]
-unknown = single_index[single_index['studyTitle'].isna()][bcl2barcode.columns]
+# Expensive operation of matching bcl2barcode index to expected Pinery index
+bcl2barcode['Index1Matched'] = bcl2barcode.apply(lambda x: find_matches(x['Index1'], x['Index1_known'], BCL2FASTQ_MISMATCH), axis=1)
+bcl2barcode['Index2Matched'] = bcl2barcode.apply(lambda x: find_matches(x['Index2'], x['Index2_known'], BCL2FASTQ_MISMATCH), axis=1)
+
+# Count how many indices were matched and get their sequence
+bcl2barcode['Index1MatchedCount'] = bcl2barcode['Index1Matched'].apply(len)
+bcl2barcode['Index2MatchedCount'] = bcl2barcode['Index2Matched'].apply(len)
+bcl2barcode['Index1MatchedSequence'] = bcl2barcode['Index1Matched'].apply(lambda x: x[0] if len(x) else '')
+bcl2barcode['Index2MatchedSequence'] = bcl2barcode['Index2Matched'].apply(lambda x: x[0] if len(x) else '')
+
+bcl2barcode["Classify"] = bcl2barcode.apply(classify, axis=1)
+
+# Sum all bcl2barcode counts that come from the same Pinery barcode
+all_known = bcl2barcode[(bcl2barcode["Classify"] == "Known")]
+all_known = all_known.groupby(
+    [PINERY_COL.SequencerRunName, PINERY_COL.LaneNumber, 'Index1MatchedSequence', 'Index2MatchedSequence', 'IndexStrategy']
```

Comment:
This is real barcode from Pinery (i.e., after 10X expansion), right?

Reply:
Correct. 10X barcode correction is one of the first steps.